Refactor IOUAmountPage

tugbadogan opened the pull request:
@trjExpensify @Julesssss @tgolen 

### Details

- Refactor IOUAmountPage as we discussed in https://github.com/Expensify/Expensify.cash/pull/2010
- Moved TextInputAutoGrow to its own component
- Moved inline styles to styles library as @marcaaron recommended.

### Fixed Issues

Fixes https://github.com/Expensify/Expensify.cash/issues/1691

### Tests
Manually tested on all platforms.

### QA Steps

- Go to the following URL: http://localhost:8080/iou/split
- Make sure you can write only decimal number(unlimited digits and up to 3 decimal numbers) to input field.
- Make sure chat compose input is working as usual.

### Tested On

- [X] Web
- [X] Mobile Web
- [X] Desktop
- [X] iOS
- [X] Android

### Screenshots

#### Web

https://user-images.githubusercontent.com/3853003/113350595-95276680-9331-11eb-8956-6c9121f424e0.mov

#### Android

https://user-images.githubusercontent.com/3853003/113350622-9e183800-9331-11eb-8de6-cedea7d9bb7c.mov

tugbadogan committed:
Merge pull request #4 from Expensify/master

Merge

tugbadogan committed:
Merge pull request #5 from Expensify/master

Merge

tugbadogan committed:
Merge pull request #6 from Expensify/master

Merge

tugbadogan committed:
Merge pull request #7 from Expensify/master

Merge

tugbadogan committed:
Merge pull request #8 from Expensify/master

Merge

tugbadogan committed:
Refactor IOUAmountPage

tugbadogan committed:
Move style object to styles

tugbadogan committed:
Merge pull request #9 from Expensify/master

Merge

marcaaron commented on src/components/TextInputAutoGrow.js:
```diff
@@ -0,0 +1,67 @@
+
+    // General styles to apply to the text input
+    // eslint-disable-next-line react/forbid-prop-types
+    style: PropTypes.any,
```
If we are making this `PropTypes.any` then we need to provide support for `PropTypes.any`. Looking at this PR and we are treating this `style` prop as an `Object`.

marcaaron replied:
Separately, having `style` and `textStyle` props is ambiguous since there's no way to know without looking at the component which refers to which. If this style is being passed to the input perhaps we should call it `inputStyle` or something more descriptive?

marcaaron commented on src/components/TextInputAutoGrow.js:
```diff
@@ -0,0 +1,67 @@
+};
+
+const defaultProps = {
+    value: undefined,
```
Looking at the one usage of this and it will always have a value. So I think we should make it required.

marcaaron commented on src/components/TextInputAutoGrow.js:
```diff
@@ -0,0 +1,67 @@
+
+const defaultProps = {
+    value: undefined,
+    style: null,
```
We should default to an empty object instead of `null` - depending on how we resolve the previous comment

marcaaron commented on src/components/TextInputAutoGrow.js:
```diff
@@ -0,0 +1,67 @@
+
+    render() {
+        const propsWithoutStyles = _.omit(this.props, ['style', 'textStyle']);
+        const widthStyle = {width: Math.max(5, this.state.textInputWidth)};
```
Generated styles should use a helper function exported from `styles/styles.js`

marcaaron commented on src/components/TextInputAutoGrow.js:
```diff
@@ -0,0 +1,67 @@
+
+    // Styles to apply to the text input
+    // eslint-disable-next-line react/forbid-prop-types
+    textStyle: PropTypes.any.isRequired,
```
Same comment about `PropTypes.any` this shouldn't be `PropTypes.any` because we are not handling the `any` case.

marcaaron commented on src/components/TextInputAutoGrow.js:
```diff
@@ -0,0 +1,67 @@
+                    {...propsWithoutStyles}
+                />
+                <Text
+                    style={[this.props.textStyle, styles.autoGrowTextInputHiddenText]}
```
What makes this "hidden text"? Could we maybe add a comment to explain what this component is supposed to be used for? It's not too clear.

marcaaron commented on src/components/TextInputFocusable/index.js:
```diff
@@ -8,7 +8,10 @@ const propTypes = {
+    defaultValue: PropTypes.string,
+
+    // The value of the comment box
+    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
```
Can we pick one type or the other? Why should we accept both `string` and `number`?

marcaaron commented on src/pages/iou/steps/IOUAmountPage.js:
```diff
@@ -65,6 +58,45 @@ class IOUAmountPage extends React.Component {
+     * @param {String} key
+     * @param {String} event
+     */
+    onlyAllowNumericValues(key, event) {
```
This function name doesn't really describe what it does and sounds more like a `boolean` value.

marcaaron commented on src/pages/iou/steps/IOUAmountPage.js:
```diff
@@ -65,6 +58,45 @@ class IOUAmountPage extends React.Component {
+     */
+    onlyAllowNumericValues(key, event) {
+        // Prevent reusing synthetic event object, so we can call preventDefault() later.
+        if (event) {
```
When would there not be an `event`?

tugbadogan replied:
We reused this function in mobile experience where we have BigNumberPad. We don't have synthetic event there. we're just passing buttonPressed key parameter.

marcaaron commented on src/pages/iou/steps/IOUAmountPage.js:
```diff
@@ -65,6 +58,45 @@ class IOUAmountPage extends React.Component {
+    onlyAllowNumericValues(key, event) {
+        // Prevent reusing synthetic event object, so we can call preventDefault() later.
+        if (event) {
+            event.persist();
```
I'm not really following why this is necessary. Can we just call `event.preventDefault()` after we check for the backspace button and the new value and then not `setState()`?

tugbadogan replied:
Actually, we don't need to call `event.preventDefault()` anymore as we started using value instead of default value.

marcaaron commented on src/pages/iou/steps/IOUAmountPage.js:
```diff
@@ -65,6 +58,45 @@ class IOUAmountPage extends React.Component {
+        }
+
+        // Backspace button is pressed
+        if (key === '<' || key === 'Backspace') {
```
What is the `<` key?

tugbadogan replied:
In BigNumberPad component that we used in mobile, we have `<` for backspace button.

marcaaron commented on src/pages/iou/steps/IOUAmountPage.js:
```diff
@@ -65,6 +58,45 @@ class IOUAmountPage extends React.Component {
+            if (this.state.amount.length > 0) {
+                this.setState(prevState => ({
+                    amount: prevState.amount.substring(0, prevState.amount.length - 1),
+                    isNextButtonEnabled: prevState.amount.length !== 1,
```
What is the significance of `!== 1`? Can we add a comment to explain what is happening here?

marcaaron commented on src/pages/iou/steps/IOUAmountPage.js:
```diff
@@ -65,6 +58,45 @@ class IOUAmountPage extends React.Component {
+        }
+        this.setState((prevState) => {
+            const newValue = `${prevState.amount}${key}`;
+            const decimalNumberRegex = new RegExp(/^\d{1,6}(\.\d{0,2})?$/, 'i');
```
What is this regex testing for? Add a comment?

tgolen replied:
I don't know if I missed this somewhere... but why is there a limit of 6 digits? Is that just ambiguous? Let's allow for any number of digits. Also with the decimals, some currencies allow three decimal places, so that needs to be updated too.

marcaaron replied:
Was your question resolved here @tgolen ?

tugbadogan replied:
Sorry, I missed this comment earlier. I changed the regex pattern with the last commit.👍

marcaaron commented on src/styles/styles.js:
```diff
@@ -1239,6 +1239,14 @@ const styles = {
+        opacity: 0,
+
+        // Using a very large value here makes sure it's out of screen
+        left: 100000,
```
I'm not really getting the hidden text thing sorry. IMO we should avoid using values like this. If we absolutely must push something off screen we can at least base this value on the screen width + current position of element.

tugbadogan committed:
Addressing comments

tugbadogan committed:
Use event.preventDefault in TextInputAutoGrow onKeyPress

This change makes input entry smoother by preventing TextInput to update its own value

tugbadogan committed:
Merge pull request #10 from Expensify/master

Merge

shawnborton commented:
Couple of design comments:
- It looks like the IOU amount is in black, but it should be in `themeColors.heading`
- It looks like the background color for the number pad buttons is missing. They should just use our default button styles.

tgolen commented on src/components/TextInputAutoGrow.js:
```diff
@@ -0,0 +1,87 @@
+                    /* eslint-disable-next-line react/jsx-props-no-spreading */
+                    {...propsWithoutStyles}
+                />
+                {this.getHiddenTextView()}
```
There shouldn't be class methods used to return things to render. Either everything should be rendered in the render method, or it should be moved to it's own functional component.

tugbadogan replied:
Thanks. I removed all functions and put everything into render function

tgolen commented on src/components/TextInputFocusable/index.js:
```diff
@@ -40,6 +43,8 @@ const propTypes = {
 };
 
 const defaultProps = {
+    defaultValue: undefined,
```
Aren't the `undefined` by default already? Maybe these should be `null` instead?

Julesssss replied:
Any reason why this can't be an empty String instead?

tugbadogan replied:
That makes sense 👍 I replaced these with empty string.

tgolen commented on src/pages/iou/steps/IOUAmountPage.js:
```diff
@@ -81,38 +104,30 @@ class IOUAmountPage extends React.Component {
+                                    }}
+                                    ref={el => this.textInput = el}
+                                    value={this.state.amount}
+                                    textAlign="left"
```
I don't see `textAlign` defined in the props. Am I just missing it somewhere? Does this need to be in props at all or is it always going to be "left"?

tugbadogan replied:
I guess I forgot to remove this prop.

tgolen reviewed (changes requested):
Looks like there are some conflicts

Julesssss commented on src/styles/styles.js:
```diff
@@ -1405,6 +1405,33 @@ function getZoomSizingStyle(isZoomed) {
+ * @param {Number} windowWidth
+ * @returns {Object}
+ */
+function getHiddenElementOutsideOfWindow(windowWidth) {
+    return {
+        position: 'absolute',
+        opacity: 0,
+        left: windowWidth,
+    };
+}
```
Are there any alternatives to this?

tugbadogan replied:
I wanted to move this hidden element out of the screen so it doesn't block any visible component. Maybe we can try using z-index and move this component background.

Julesssss replied:
I'd be interested in seeing if a negative z-index gives us a slightly simpler solution, but I don't think it should block the PR.

marcaaron replied:
there is also `transform: translateX()` as an option

tugbadogan replied:
I used transform solution to hide the element like this:

```
hiddenElementOutsideOfWindow: {
        position: 'fixed',
        top: 0,
        left: 0,
        opacity: 0,
        transform: 'translateX(-100%)',
    }, 
```

tugbadogan committed:
Merge pull request #11 from Expensify/master

Merge

tugbadogan committed:
Merge pull request #12 from Expensify/master

Merge

tugbadogan commented:
@shawnborton

> Couple of design comments:
> 
> * It looks like the IOU amount is in black, but it should be in `themeColors.heading`

I updated IOU Amount color. 

> * It looks like the background color for the number pad buttons is missing. They should just use our default button styles.

I'm already using `styles.button` style for number pad buttons. Am I missing something here? They look like design screenshots in the issue.

https://github.com/Expensify/Expensify.cash/blob/d3f1d28478357e2a66b155e660def191d6ed79a8/src/components/BigNumberPad.js#L49-L53

shawnborton commented:
From the video you posted, it looks like the buttons do not have a background color:
![image](https://user-images.githubusercontent.com/2319350/114428728-6602c780-9b71-11eb-9d83-7e1600e0f86c.png)


But perhaps that's just the video not showing all of the colors.

tugbadogan committed:
Addressing comments

tgolen commented:
Can you please fix the conflicts before I take another look at this?

tugbadogan committed:
Merge pull request #13 from Expensify/master

Merge

tugbadogan committed:
Merge branch 'master' into tugbadogan-refactor-iou-amount-page

tugbadogan commented:
> Can you please fix the conflicts before I take another look at this?

@tgolen 
I resolved the conflicts. Can you review again?

marcaaron commented on src/components/TextInputAutoGrow.js:
```diff
@@ -0,0 +1,80 @@
+import {View, Text} from 'react-native';
+import PropTypes from 'prop-types';
+import _ from 'underscore';
+import {getAutoGrowTextInputStyle, getHiddenElementOutsideOfWindow} from '../styles/styles';
```
This should be `getHiddenElementOutsideOfWindowStyle()` to match other style helper methods

marcaaron commented on src/components/TextInputAutoGrow.js:
```diff
@@ -0,0 +1,80 @@
+    inputStyle: {},
+};
+
+class TextInputAutoGrow extends React.Component {
```
I feel like this component has a confusing name since the `TextInputFocusable` component does "auto grow" by setting `numLines`. This component looks like more of a "variable width text input", but "auto grow" is ambigous.

Not sure if others feel the same way. I wasn't too sure why we need this component or what it does until digging into the logic.

tugbadogan replied:
I renamed it as `TextInputAutoWidth`. I think the component's behaviour is more understandable now.

marcaaron reviewed (changes requested):
Mostly looks good. Only have some small comments about naming conventions.

tugbadogan committed:
Addressing comments

tugbadogan committed:
Updating regex comment

tugbadogan commented:
@tgolen I updated the PR. Can you review again?

Julesssss commented on src/styles/styles.js:
```diff
@@ -1266,6 +1266,14 @@ const styles = {
         zIndex: 10,
     },
+
+    hiddenElementOutsideOfWindow: {
```
This should be updated too I think, as we're no longer hiding it outside of the window.

tugbadogan replied:
I wanted to keep the "OutsideOfWindow" part as this element is still outside of the window (positioned right outside of the top/left corner).

Julesssss replied:
Ah, I see. Thanks for pointnng that out.

marcaaron commented:
I believe this PR may have caused a regression related to many text inputs. 
In some cases, they become completely unresponsive... cc @stitesExpensify

marcaaron commented:
The problem is that if we have no `value` or `defaultValue` passed then it will pass a prop of `value: ""` to a `TextInput`

tugbadogan commented:
Thanks for the flag @marcaaron , creating a PR now.

isagoico commented:
I don't think we're able to QA this PR since it requires changing some code. Can you guys assign someone to QA and let me know once it's done so we can check it off the list? Thanks in advance. CC @marcaaron @tgolen @Julesssss @ctkochan22

marcaaron commented:
@isagoico Seems testable by using a deep link or entering url in web? Which code changes are you referring to?

isagoico commented:
Oh idk why I thought there was a code change needed here 🤦 my apologies. 

>Go to the following URL: http://localhost:8080/iou/split

Can you share the deeplink URL that we can use to test this?

marcaaron commented:
I believe `https://staging.cash/iou/split` should work.

isagoico commented:
I think that worked. 
![image](https://user-images.githubusercontent.com/44479856/114772494-de1ce900-9d3b-11eb-813f-d0d1c1b6c0da.png)

Should we test this only on web and mWeb? Not sure if we're able to navigate to that URL or feature in the apps (android, iOS, dekstop)

marcaaron commented:
It _should_ work on all platforms except desktop I think?

isagoico commented:
## Split - Able to enter more than 8 decimals and LHN/conversation area is blank 

## Expected Result
Make sure you can write only decimal number(max 6 digit and 2 decimal) to input field.

## Actual Result
Able to enter 3 decimal inputs

## Action performed
1. Launch the url and login
2. Add iou/split to the URL
3. Input decimal in the field

## Platform
Issue is confirmed in: 

Android ✔️ 
Web ✔️ 

**Build version:** 1.0.21-0
 
 **Notes/Images/Video**
![image](https://user-images.githubusercontent.com/44479856/114788490-8d16f000-9d4f-11eb-8e82-817caddc602a.png)

marcaaron commented:
I think this test step is out of date and needed to be updated.

tugbadogan commented:
Sorry, I forgot to update the test plan. Fixed it now.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/Expensify.cash
    /actions/runs/765130740) 🚀 to 
    staging on Mon Apr 19 2021 at 22:41:56 GMT+0000 (Coordinated Universal Time)

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
